scripts/apps/authoring/authoring/components/CharacterCountConfigButton.tsx

```diff
@@ -0,0 +1,150 @@
+import {gettext} from 'core/utils';
+import {ModalBody} from 'core/ui/components/Modal/ModalBody';
+import {ModalFooter} from 'core/ui/components/Modal/ModalFooter';
+import {Radio, CheckGroup} from 'superdesk-ui-framework';
+import ng from 'core/services/ng';
+import {dispatchInternalEvent} from 'core/internal-events';
+
+export const CHARACTER_LIMIT_UI_PREF = 'editor:char_limit_ui';
+
+export type CharacterLimitUiBehavior = 'highlight' | 'limit'; // highlight extra chars or limit the editor
+export interface ICharacterLimitUiBehavior {
+    [schemaField: string]: CharacterLimitUiBehavior;
+}
+
+export const DEFAULT_UI_FOR_EDITOR_LIMIT: CharacterLimitUiBehavior = 'highlight';
+
+interface IProps {
+    field: string;
+}
+
+interface IState {
+    preferences: any;
+}
+
+export class CharacterCountConfigButton extends React.Component<
+    IProps,
+    IState
+> {
+    preferencesService: any;
+
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            preferences: null,
+        };
+
+        this.preferencesService = ng.get('preferencesService');
+        this.onModalValueChange = this.onModalValueChange.bind(this);
+    }
+
+    componentDidMount() {
+        this.preferencesService.get().then((preferences) => {
+            this.setState({
+                preferences,
+            });
+        });
+    }
+
+    onModalValueChange(newValue: CharacterLimitUiBehavior) {
+        if (!newValue) {
+            return;
+        }
+
+        const newPreferences = {
+            ...this.state.preferences,
+            [CHARACTER_LIMIT_UI_PREF]: {
+                ...this.state.preferences[CHARACTER_LIMIT_UI_PREF],
+                [this.props.field]: newValue,
+            },
+        };
+
+        this.setState({preferences: newPreferences});
+        this.preferencesService.update(newPreferences);
+        dispatchInternalEvent('changeUserPreferences', newPreferences);
+    }
+
+    render() {
+        if (this.state.preferences == null) {
+            return null;
+        }
+
+        return (
+            <button
+                className="char-count-config-button"
+                onClick={() => {
+                    showModal((props) => (
+                        <CharacterCountConfigModal
+                            closeModal={props.closeModal}
+                            value={
+                                this.state.preferences[
+                                    CHARACTER_LIMIT_UI_PREF
+                                ]?.[this.props.field] ?? DEFAULT_UI_FOR_EDITOR_LIMIT
+                            }
+                            onChange={this.onModalValueChange}
+                        />
+                    ));
+                }}
+            >
+                <i className="icon-settings" />
+            </button>
+        );
+    }
+}
+
+function CharacterCountConfigModal({closeModal, onChange, value}) {
+    const [radioValue, radioValueSet] = React.useState<
```

Comment:
Still trying to sneak hooks in ;D My bad, I didn't follow through with a lint rule :smile:

Reply:
hahaha I swear I didn't mean to sneak it, I'm just used to it. Anyway they're pretty straightforward and they're becoming the standard in React, we can talk about it with other devs if you want

Reply:
I don't think there is a standard here, just two APIs - hooks or `setState` - both work and have their own pros/cons.

All the components in this project use `setState` / redux so I wouldn't like to bring in another way to manage state if it doesn't provide significant benefits. This component is very basic and doesn't do anything so I'd keep it as a `PureComponent`.

Reply:
Hooks are meant to replace class-based state. In React docs they say that it's the recommended way of creating new components. I never meant to replace the current classes we have, unless we create custom hooks, which in the future could be a very powerful way of handling CRUD operations, like on this example:

```tsx
const UserComponent = ({ id }) => {
  const [user, editUser, removeUser] = userApi.useUser(id) // custom hook that connects an API with a UI component
  return <div>{user?.name}</div>
}
```

I don't mean that it cannot be made with class-logic but IMO it's a cleaner way of connecting the data to the UI, smaller and more intuitive to read.

Also, setState is for component state and redux for app state, that we only use on editor3 for now, so I don't see an issue with hooks for component state.

Reply:
> Hooks are meant to replace class-based state. In React docs they say that it's the recommended way of creating new components.

Class components aren't deprecated and there are no plans to do so. Given it's a core API I doubt it will ever be removed.

I don't mind hooks in principle. If all our code was in hooks and you wanted to add a class component, I would ask you not to. I want the project to use as little different patterns as possible so there is less to learn in order to maintain it. Hooks aren't not just an addition, as a library would be, but a totally different paradigm. Everyone would need to learn how all life-cycle methods are implemented with hooks and IMO we could spend that time more effectively refactoring business logic.

Reply:
You're kinda right in the sense that it's better to keep it uniform but getting stuck in a way a library works instead of exploring the future of it (classes won't be deprecated soon, but all the improvements will come from hooks logic) is not great either. If we work with react it's reasonable to keep learning it and keep moving with the library and the community.

Reply:
I get your point. How about we keep class components in the core and build a next few extensions using hooks to try it out?

Reply:
hmmm you mean moving component to a new extension?

Reply:
No :) Just using a class component here and when we have to make a new extension build it all with hooks.